### PR TITLE
add case for virtio memory with various memory allocation and guest numa

### PR DESCRIPTION
**Pull request by nanli1**
xxxx-299031 Virtio-mem device with various memory allocation and guest numa settings.

```

 avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 memory.devices.virtio_mem.memory_allocation_and_numa
 (01/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_with_node.without_numa.no_maxmemory: PASS (40.78 s)
 (02/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_with_node.without_numa.with_maxmemory: PASS (75.51 s)
 (03/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_with_node.with_numa.no_maxmemory: PASS (42.39 s)
 (04/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_with_node.with_numa.with_maxmemory: PASS (75.50 s)
 (05/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_without_node.without_numa.no_maxmemory: PASS (40.85 s)
 (06/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_without_node.without_numa.with_maxmemory: PASS (76.11 s)
 (07/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_without_node.with_numa.no_maxmemory: PASS (42.37 s)
 (08/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_without_node.with_numa.with_maxmemory: PASS (40.88 s)
 (09/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.define_guest.virtio_mem_with_exceed_size.without_numa.with_maxmemory: PASS (40.75 s)
 (10/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_with_node.without_numa.no_maxmemory: PASS (42.51 s)
 (11/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_with_node.without_numa.with_maxmemory: PASS (43.75 s)
 (12/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_with_node.with_numa.no_maxmemory: PASS (42.01 s)
 (13/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_with_node.with_numa.with_maxmemory: PASS (43.59 s)
 (14/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_without_node.without_numa.no_maxmemory: PASS (41.96 s)
 (15/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_without_node.without_numa.with_maxmemory: PASS (42.07 s)
 (16/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_without_node.with_numa.no_maxmemory: PASS (40.25 s)
 (17/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_without_node.with_numa.with_maxmemory: PASS (40.28 s)
 (18/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.cold_plug.virtio_mem_with_exceed_size.without_numa.with_maxmemory: PASS (42.23 s)
 (19/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_with_node.without_numa.no_maxmemory: PASS (77.94 s)
 (20/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_with_node.without_numa.with_maxmemory: PASS (77.19 s)
 (21/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_with_node.with_numa.no_maxmemory: PASS (77.04 s)
 (22/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_with_node.with_numa.with_maxmemory: PASS (77.42 s)
 (23/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_without_node.without_numa.no_maxmemory: PASS (75.06 s)
 (24/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_without_node.without_numa.with_maxmemory: PASS (76.89 s)
 (25/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_without_node.with_numa.no_maxmemory: PASS (75.52 s)
 (26/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_without_node.with_numa.with_maxmemory: PASS (73.62 s)
 (27/27) type_specific.io-github-autotest-libvirt.memory.devices.virtio_mem.memory_allocation_and_numa.hot_plug.virtio_mem_with_exceed_size.without_numa.with_maxmemory: PASS (77.26 s)

```

### CODE REVIEW

**Comment by nanli1** [comment]
@dzhengfy Please review virtio-mem case ,Thank you

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg [hunk 1]
```diff
@@ -0,0 +1,81 @@
+- memory.devices.virtio_mem.memory_allocation_and_numa:
+    type = virtio_mem_with_memory_allocation_and_numa
+    no s390-virtio
+    start_vm = no
+    mem_model = "virtio-mem"
+    mem_value = 2097152
```
Comment (dzhengfy) [inline libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg:6]: Please increase to 8G

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    current_mem = 2097152
```
Comment (dzhengfy) [inline libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg:7]: Please increase to 8G

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            max_attrs = ""
+            define_error = "${err_msg1}"
+        - with_maxmemory:
+            max_mem = 10485760
+            max_attrs = '"max_mem_rt": ${max_mem}, "max_mem_rt_unit": "KiB",'
+    variants numa_topology:
+        - without_numa:
+            numa_attrs = ""
+        - with_numa:
+            numa_mem = 2097152
```
Comment (dzhengfy) [inline libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg:32]: 8G

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            with_maxmemory:
+                hot_plug_error = "${err_msg2}"
+                cold_plug_error = "${err_msg2}"
+            no_maxmemory:
+                cold_plug_error = "${err_msg1}"
+                without_numa,with_numa:
+                    hot_plug_error = "${err_msg1}"
+        - virtio_mem_with_exceed_size:
+            only without_numa.with_maxmemory
+            target_size = "3145728"
```
Comment (dzhengfy) [inline libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg:63]: we need increase this size to make sure exceed size

File: libvirt/tests/src/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            expected_error=cold_plug_error, flagstr=plug_option)
+        if cold_plug_error:
+            return
+
+        test.log.info("TEST_STEP3: Start guest and check result")
+        res = virsh.start(vm_name)
+        if coldplug_start_error:
+            libvirt.check_result(res, expected_fails=coldplug_start_error)
+        else:
+            libvirt.check_exit_status(res)
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.py:145]: line 142 to 145 should be replaced with single line, right?
` libvirt.check_result(res, expected_fails=coldplug_start_error)`

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg (hunk 1 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg]: Thanks dzheng, Updated

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg (hunk 2 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg]: Thanks dzheng, Updated

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg (hunk 3 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg]: Thanks dzheng, Updated

File: libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg (hunk 4 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.cfg]: Thanks dzheng, Updated

File: libvirt/tests/src/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.py (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_devices/virtio_mem_with_memory_allocation_and_numa.py]: Thanks dzheng, Updated

**Comment by nanli1** [comment]
@dzhengfy Could you please re-review

**Review by dzhengfy — APPROVED** [review]
LGTM

**Comment by liang-cong-red-hat** [comment]
LGTM

**Pull request merged**